src/registry.ts

```diff
@@ -0,0 +1,67 @@
+/**
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {UpgradePass, UpgradePassCollection} from './upgrade-pass';
+
+/**
+ * A singleton class where upgrade passes (and collections thereof) can register
+ * themselves and you can get a collection of passes by querying.
+ */
+export class UpgradePassRegistry {
+  private _all = new Map<string, UpgradePass<any>|UpgradePassCollection>();
+
+  private constructor() {}
+
+  register(pass: UpgradePass<any>|UpgradePassCollection) {
+    this._all.set(pass.code, pass);
+  }
+
+  /**
+   * Given an array of string codes for registered passes and pass collections,
+   * return the set of passes.
+   *
+   * Treats the code 'all' special and just returns every registered upgrade
+   * pass.
+   */
+  getPasses(passCodes: string[]): Set<UpgradePass<any>> {
+    const results = new Set<UpgradePass<any>>();
+    for (const code of passCodes) {
```

Comment:
Why require that `all` must be alone?

If we require that UpgradePassCollections be pre-flattened then you run into some potentially weird timing-related issues when defining a collection, where you need to define your collections in the right order if they reference each other.

How about this decomposition? I also added cycle handling.